rest_tools/client/client.py

```diff
@@ -40,35 +41,40 @@ class RestClient:
         password (str): (optional) auth-basic password
     """
 
-    def __init__(self, address, token=None, timeout=60.0, retries=10, **kwargs):
+    def __init__(
+        self,
+        address: str,
+        token: Optional[Union[str, bytes, Callable[[], Union[str, bytes]]]] = None,
+        timeout: float = 60.0,
+        retries: int = 10,
+        **kwargs: Any
+    ) -> None:
         self.address = address
         self.timeout = timeout
         self.retries = retries
         self.kwargs = kwargs
-        self.session = None
-
         self.logger = logging.getLogger('RestClient')
         self.logger.setLevel('DEBUG')
 
         # token handling
-        self.access_token = None
-        self.token_func = None
+        self.access_token: Optional[Union[str, bytes]] = None
+        self.token_func: Optional[Callable[[], Union[str, bytes]]] = None
         if token:
             if isinstance(token, (str,bytes)):
                 self.access_token = token
             elif callable(token):
                 self.token_func = token
 
-        self.open() # start session
+        self.session = self.open()  # start session
```

Comment:
It seems odd to have the session assigned here, yet it's also assigned inside `open()`.

Reply:
This is an odd scenario. Before in `__init__()` we assigned `self.session = None` but then immediately reassign it in `open()` -- it also leads to a typing nightmare downstream because `self.session` is presumed to be `None`.

On the other hand, declaring instance variables outside of `__init__` is bad form.

So this was the compromise I found.